adr/keycloak-adr.md

```diff
@@ -1,264 +1,139 @@
-
-#### The Update Framework (TUF) and go-tuf
-
-- **go-tuf (The Update Framework in Go)**: For insights into secure distribution and integrity verification mechanisms, go-tuf implements The Update Framework (TUF) in Go. TUF is designed to prevent various attacks on software update systems and can inspire approaches to ensuring the integrity of immutable logs with hash chaining. More information can be found in the [go-tuf GitHub repository](https://github.com/theupdateframework/go-tuf).
-
-#### nats.io references
-- [NATS Documentation](https://docs.nats.io/)
-- [NATS on Kubernetes](https://docs.nats.io/nats-on-kubernetes/nats-on-kubernetes)
-- [NATS Security](https://docs.nats.io/nats-on-kubernetes/nats-on-kubernetes)
+#### SAML Integration
+**Pros:**
+- Lets you automatically sync user groups between Keycloak and GitLab, making access control more accessible.
+
+**Cons:**
+- Setting it up can be complicated, so careful alignment between Keycloak and GitLab is needed.
+- To use all features, you might need a higher tier of GitLab, which could mean higher costs.
+- If you have a lot of users, you might need to pay for more expensive versions of Keycloak, which could be a bit of a hassle.
 
+### Provision Users
+Keycloak will provision users and manage user roles, groups, and permissions. This ensures user access is managed centrally and consistently across all applications and services.
 
+#### SCIM Integration
+**Pros:**
+- Easy to Mix Systems: SCIM is a common framework for sharing user info between different programs or online services, making them work smoothly.
+- Saves Time: It handles adding, updating, and removing user accounts by itself, which means less work for people managing these tasks.
+- Fewer Mistakes: Since it's automated, there's less chance someone will type something wrong or forget to update important info.
+
+**Cons:**
+- Can Be Tricky to Set Up: Getting SCIM to work just right with all your systems can be complicated.
+- Learning Curve: If you're not familiar with SCIM, it might take a while to get the hang of it.
+
+#### Manual Provisioning
+**Pros:**
+- Full control over the user provisioning process.
+- Can be implemented without additional development or integration efforts.
+
+**Cons:**
+- Time-consuming and not scalable.
+- Prone to human error, potentially leading to security and compliance issues.
+
+#### Custom Code
+**Pros:**
+- Flexibility: You can tailor the provisioning process to your exact needs, ensuring it fits your organization perfectly.
+- No exteranl depdency: You can do this with what GitLab already offers, no need for external OSS tools.
+
+**Cons:**
+- More Work: You'll have to write and maintain the code yourself, which can be a lot of work.
+
+### Deprovision Users
+Keycloak will be used to de-provision users and manage user access rights, ensuring timely revocation of access when users leave the organization or change roles.
```

Comment:
```suggestion
Users need to be deprovisioned (GitLab user account disabled or deleted) promptly from GitLab when they lose access via Keycloak (Keycloak user account disabled/deleted or removed from a group that gives GitLab access).

Note: it possible to authenticate to GitLab with Personal Access Tokens (PATs) and SSH keys, these bypass SSO because the user does not login with a browser. It is critical to promptly deactivate or delete users from GitLab to terminate access.

Note: deleted and deactivated users do not consume a license. GitLab bills per user active in the past 30 days(?). To minimize cost, users should be deactivated promptly.
```